### PR TITLE
refactor(x/ecocredit): update market params API

### PR DESCRIPTION
**Pull request by technicallyty**
## Description

- updates the marketplace params proto definitions

Closes: n/a

---

### Author Checklist

*All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.*

I have...

- [ ] included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] added `!` to the type prefix if API or client breaking change
- [ ] targeted the correct branch (see [PR Targeting](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#pr-targeting))
- [ ] provided a link to the relevant issue or specification
- [ ] followed the guidelines for [building modules](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules)
- [ ] included the necessary unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#testing)
- [ ] added a changelog entry to `CHANGELOG.md`
- [ ] included comments for [documenting Go code](https://blog.golang.org/godoc)
- [ ] updated the relevant documentation or specification
- [ ] reviewed "Files changed" and left comments if necessary
- [ ] confirmed all CI checks have passed

### Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items.*

I have...

- [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] confirmed `!` in the type prefix if API or client breaking change
- [ ] confirmed all author checklist items have been addressed 
- [ ] reviewed state machine logic
- [ ] reviewed API design and naming
- [ ] reviewed documentation is accurate
- [ ] reviewed tests and test coverage
- [ ] manually tested (if applicable)

### CODE REVIEW

File: proto/regen/ecocredit/marketplace/v1/tx.proto [hunk 1]
```diff
@@ -25,8 +25,8 @@ service Msg {
   // Buy creates credit buy orders.
   rpc Buy(MsgBuy) returns (MsgBuyResponse);
 
-  // AllowAskDenom is a governance operation which authorizes a new ask denom to
-  // be used in sell orders
+  // AllowAskDenom is a governance operation which authorizes new ask denoms and/or removes denoms to
+  // be used in sell orders.
   rpc AllowAskDenom(MsgAllowAskDenom) returns (MsgAllowAskDenomResponse);
```
Comment (ryanchristo) [inline proto/regen/ecocredit/marketplace/v1/tx.proto:30]: I think the original intention here was to add an ask denom, i.e. "to allow it". We should probably have two separate methods, one for adding and one for removing and we should only allow one at a time. I think we should avoid batch updates to governance params and each addition/removal warrants its own governance proposal.

**Review by ryanchristo — COMMENTED** [review]
As mentioned in #960 and #962, it looks like we may not update to Cosmos SDK v0.46 for Regen Ledger v4.0. We can still add the proto updates and the message implementations but we will need to find a workaround for allowed ask denoms until the v0.46 update.

In line with my comments in those two pull requests, I think we should avoid batch updates for governance methods as each addition/removal should warrant its own proposal.

File: proto/regen/ecocredit/marketplace/v1/tx.proto (hunk 1 above)
Comment (technicallyty) [reply proto/regen/ecocredit/marketplace/v1/tx.proto:30]: yeah i think it makes sense to remove batching. governance can do batch updates if needed anyway, as the new v0.46 gov updates allows for multiple messages to execute from a single proposal

File: proto/regen/ecocredit/marketplace/v1/tx.proto [hunk 2]
```diff
@@ -195,21 +195,32 @@ message MsgBuyResponse {
...
 
+
 // MsgAllowAskDenom is the Msg/AllowAskDenom request type.
 message MsgAllowAskDenom {
-  // root_address is the address of the governance account which can authorize
-  // ask denoms
+
+  // root_address is the address of the signer.
+  // This MUST equal the address of the gov module for the tx to succeed.
   string root_address = 1;
```
Comment (robert-zaremba) [inline proto/regen/ecocredit/marketplace/v1/tx.proto:204]: root sounds ambiguous. Why not `sender` or `signer`?

File: proto/regen/ecocredit/marketplace/v1/tx.proto (hunk 1 above)
Comment (robert-zaremba) [reply proto/regen/ecocredit/marketplace/v1/tx.proto:30]: TBH, I like the idea of packing this into one proposal (as it's done now). Imagine there are 10 denoms, and you want to change them to 10 other denoms. That would require 20 proposals. 
At the same time, if there was no clarity during the signaling, one can still do small changes using a proposal with only one edit.

File: proto/regen/ecocredit/marketplace/v1/tx.proto [hunk 3]
```diff
@@ -195,21 +195,32 @@ message MsgBuyResponse {
...
 
-  // denom is the denom to allow (ex. ibc/GLKHDSG423SGS)
-  string denom = 2;
+  // add_denoms specifies denoms to add to the allowed denom list
```
Comment (robert-zaremba) [inline proto/regen/ecocredit/marketplace/v1/tx.proto:206]:
```suggestion
  // add_denoms specifies denoms to add to the allowed denom list.
  // It's OK to include here a denom which is already allowed. 
  // Returns an error if a union of `add_denom` and `remove_denom` is not empty.
```

File: proto/regen/ecocredit/marketplace/v1/tx.proto [hunk 4]
```diff
@@ -195,21 +195,32 @@ message MsgBuyResponse {
...
+  repeated DenomInfo add_denoms = 2;
+
+  // remove_denoms specifies denoms to remove from the allowed denom list
```
Comment (robert-zaremba) [inline proto/regen/ecocredit/marketplace/v1/tx.proto:209]:
```suggestion
  // remove_denoms specifies denoms to remove from the allowed denom list.
  // Returns an error if a union of `add_denom` and `remove_denom` is not empty.
```

File: proto/regen/ecocredit/marketplace/v1/tx.proto [hunk 5]
```diff
@@ -195,21 +195,32 @@ message MsgBuyResponse {
...
+  // DenomInfo contains the information related to a bank denom
+  message DenomInfo {
+    // denom is the denom to allow (ex. ibc/GLKHDSG423SGS)
+    string denom = 1;
 
-  // exponent is the exponent that relates the denom to the display_denom and is
-  // informational
-  uint32 exponent = 4;
+    // display_denom is the denom to display to the user and is informational
+    string display_denom = 2;
```
Comment (robert-zaremba) [inline proto/regen/ecocredit/marketplace/v1/tx.proto:218]: not sure why this is needed? I think we should use the default display from the bank module.

File: proto/regen/ecocredit/marketplace/v1/tx.proto [hunk 6]
```diff
@@ -195,21 +195,32 @@ message MsgBuyResponse {
...
+
+    // exponent is the exponent that relates the denom to the display_denom and
+    // is informational
+    uint32 exponent = 3;
```
Comment (robert-zaremba) [inline proto/regen/ecocredit/marketplace/v1/tx.proto:222]: again, why denormalizing and not using the information from bank?

**Review by robert-zaremba — COMMENTED** [review]
I would prefer that we use information from the bank instead of denormalizing it in the `allow denom list`

**Comment by ryanchristo** [comment]
Closing per #1342 and #1346

**Pull request closed without merging**